contracts/lottery/src/test.rs

```diff
@@ -0,0 +1,786 @@
+    tickets.set(Address::random(&env), vec![&env, vec![&env, 22, 14, 35, 1, 2]]);
+    tickets.set(Address::random(&env), vec![&env, vec![&env, 22, 14, 3, 1, 2]]);
+
+    // there are:
+    // - 4 tickets with 5 hits
+    // - 2 ticket with 4
+    // - 1 ticket with 3
+    // that gives a total of 160% of pool balance in prizes
+
+
+    let winners = get_winners(&env, &result, &tickets, &thresholds);
+    let total_prizes_percentage = count_total_prizes_percentage(&winners, &thresholds);
+    recalculate_new_thresholds(&winners, &mut thresholds, total_prizes_percentage);
+    assert_eq!(3, thresholds.len());
+    assert_eq!(6, thresholds.get(3).unwrap());
+    assert_eq!(9, thresholds.get(4).unwrap());
+    assert_eq!(18, thresholds.get(5).unwrap());
+}
+
+#[test]
+fn prizes_are_properly_calculated_and_assigned() {
+    let env = Env::default();
+    let p1 = Address::random(&env);
+    let p2 = Address::random(&env);
+    let p3 = Address::random(&env);
+
+    let (result, tickets, mut thresholds) = setup_addtional_test_data(&env, p1.clone(), p2.clone(), p3.clone());
+    let winners = get_winners(&env, &result, &tickets, &thresholds);
+
+    let pool = 400;
+    let prizes = calculate_prizes(&env, &winners, &mut thresholds, pool);
+
+    assert_eq!(3, prizes.len());
+    assert_eq!(112, prizes.get(p1.clone()).unwrap());
+    assert_eq!(112, prizes.get(p2.clone()).unwrap());
+    assert_eq!(168, prizes.get(p3.clone()).unwrap());
+}
+
+fn assert_auth(
+    auths: &[(Address, AuthorizedInvocation)],
+    idx: usize,
+    call_addr: Address,
+    auth_addr: Address,
+    func: Symbol,
+    args: Vec<Val>,
+) {
+    let auth = auths.get(idx).unwrap();
+    assert_eq!(auth.0, call_addr);
+    assert_eq!(
+        auth.1.function,
+        AuthorizedFunction::Contract((auth_addr, func, args))
+    );
+}
+
+struct TestScenario<'a> {
+    env: Env,
+    contract_id: Address,
+    client: LotteryContractClient<'a>,
+    test_token_client: token::StellarAssetClient<'a>,
+    token_client: token::Client<'a>
+}
+
+fn setup_test_scenario<'a>() -> TestScenario<'a> {
+        
+    let env = Env::default();
+    env.mock_all_auths();
+    let contract_id = env.register_contract(None, LotteryContract);
+    let client = LotteryContractClient::new(&env, &contract_id);
+    let token_admin = Address::random(&env);
+    let token_addr = env.register_stellar_asset_contract(token_admin.clone());
+    let test_token_client = token::StellarAssetClient::new(&env, &token_addr);
+    let token_client = token::Client::new(&env, &token_addr);
+
+    TestScenario { 
+        env, 
+        contract_id, 
+        client,
+        test_token_client, 
+        token_client
+    }
+}
+
+fn setup_addtional_test_data(
+    env: &Env,
+    add1: Address,
+    add2: Address,
+    add3: Address,
+) -> (
+    LotteryResult,
+    Map<Address, Vec<LotteryTicket>>,
+    Map<u32, u32>,
+) {
+    let result = draw_numbers(&env, 50, 5, 666);
```

Comment:
I have some concerns of using a production code function for generating data for other tests. Would it be better to get a result and hardcode that `Vec` ? So we ensure a bug in `draw_numbers` is not promoting false positives.

Reply:
Also, clearer for reviewing tests :P